lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::{HashMap, BTreeSet};
+use std::cell::RefCell;
+use std::rc::Rc;
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+use itertools::Itertools;
+
+enum Action {
+    KeyUp,
+    KeyDown
+}
+
+type Bindings = HashMap<String, Box<dyn FnMut()>>;
+
+#[wasm_bindgen]
+pub struct KeyboardEngine {
+    bindings:        Rc<RefCell<Bindings>>,
+    _element:         web_sys::Element,
+    _prevent_default: bool,
+    //keydown_cb:    Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    //keyup_cb:      Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    _state:          Rc<RefCell<BTreeSet<String>>>
+}
+
+impl KeyboardEngine {
+    pub fn new(element: &web_sys::Element, prevent_default: bool) -> Self {
```

Comment:
Using `bool` in arguments is almost always a bad idea. When you see code `KeyboardEngine::new(foo, true)` you don't understand what this `true` means. It is always better either to create custom variants or use the builder pattern / modifier, like `KeyboardEngine::new("foo").prevent_default()` instead. 😃